README.md
# 1code-rust
Rust version of https://github.com/soramitsu/1code-java

## Features 

Serialization and Deserialization for [1code format](https://github.com/soramitsu/1code-java/blob/master/1code-spec.md)
all structs with `serde::Serializer` and/or `serde::Deserializer` traits.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Serialization functionality with a test case (all possible data types but not all possible data ranges). 
- Deserialization functionality with a test case (all possible data types except enums but not all possible data ranges). 
- Benchmarks to compare with future versions and alternative implementatons and formats.


benches/tests.rs
#[macro_use]
extern crate bencher;

use bencher::Bencher;
use onecode::ser as one_code_ser;

use std::collections::HashMap;
#[derive(serde::Serialize)]
struct StructToSerialize {
    boolean: bool,
    positive_integer: u8,
    negative_integer: i16,
    positive_float: f32,
    negative_float: f32,
    negative_float_comma: f32,
    empty_string: String,
    number_string: String,
    latin_string: String,
    cyrillic_string: String,
    japan_string: String,
    number_list: Vec<u16>,
    string_list: Vec<String>,
    number_dictionary: HashMap<String, u16>,
    string_dictionary: HashMap<String, String>,
    list_dictionary: HashMap<String, Vec<u16>>,
    null: Option<()>,
}

fn serialize(bench: &mut Bencher) {
    let mut number_dictionary = HashMap::new();
    number_dictionary.insert("1".to_string(), 1);
    let mut string_dictionary = HashMap::new();
    string_dictionary.insert("1".to_string(), "1".to_string());
    let mut list_dictionary = HashMap::new();
    list_dictionary.insert("1".to_string(), vec![1, 2, 3]);
    let struct_to_serialize = StructToSerialize {
        boolean: true,
        positive_integer: 1,
        negative_integer: -1,
        positive_float: 1.5,
        negative_float: -1.5,
        negative_float_comma: -1.5,
        empty_string: "".to_string(),
        number_string: "0.1".to_string(),
        latin_string: "hello world".to_string(),
        cyrillic_string: "привет мир".to_string(),
        japan_string: "こんにちは世界".to_string(),
        number_list: vec![1, 2, 3],
        string_list: vec!["1".to_string(), "01".to_string(), "011".to_string()],
        number_dictionary: number_dictionary,
        string_dictionary: string_dictionary,
        list_dictionary: list_dictionary,
        null: Option::None,
    };
    bench.iter(|| {
        one_code_ser::to_string(&struct_to_serialize).unwrap();
    });
}

benchmark_group!(benches, serialize);
benchmark_main!(benches);


src/de.rs
use std::ops::{AddAssign, MulAssign, Neg};

use serde::de::{self, DeserializeSeed, EnumAccess, MapAccess, SeqAccess, VariantAccess, Visitor};
use serde::Deserialize;

use crate::error::{Error, Result};

pub struct Deserializer<'de> {
    input: &'de str,
}

impl<'de> Deserializer<'de> {
    pub fn from_str(input: &'de str) -> Self {
        Deserializer { input }
    }
}

pub fn from_str<'a, T>(s: &'a str) -> Result<T>
where
    T: Deserialize<'a>,
{
    let mut deserializer = Deserializer::from_str(s);
    let t = T::deserialize(&mut deserializer)?;
    if deserializer.input.is_empty() {
        Ok(t)
    } else {
        Err(Error::TrailingCharacters)
    }
}

impl<'de> Deserializer<'de> {
    /// Look at the first character in the input without consuming it.
    fn peek_char(&mut self) -> Result<char> {
        self.input.chars().next().ok_or(Error::Eof)
    }

    /// Consume the first character in the input.
    fn next_char(&mut self) -> Result<char> {
        let ch = self.peek_char()?;
        self.input = &self.input[ch.len_utf8()..];
        Ok(ch)
    }

    /// | rust  | 1coded |
    /// | ----- | ------ |
    /// | true  |   T    |
    /// | false |   F    |
    fn parse_bool(&mut self) -> Result<bool> {
        if self.input.starts_with("T") {
            self.input = &self.input["T".len()..];
            Ok(true)
        } else if self.input.starts_with("F") {
            self.input = &self.input["F".len()..];
            Ok(false)
        } else {
            Err(Error::ExpectedBoolean)
        }
    }

    /// Leading zeros aren't allowed in `1code`.
    /// | rust  | 1coded |
    /// | ----- | ------ |
    /// | 0     | i0e    |
    /// | 42    | i42e   |
    /// | -1    | i-1e   |
    /// | 1.5   | i1.5e  |
    fn parse_unsigned<T>(&mut self) -> Result<T>
    where
        T: AddAssign<T> + MulAssign<T> + From<u8>,
    {
        if self.next_char()? != 'i' {
            return Err(Error::ExpectedInteger);
        }
        let mut int = T::from(0);
        loop {
            match self.input.chars().next() {
                Some(ch @ '0'..='9') => {
                    self.input = &self.input[1..];
                    int *= T::from(10);
                    int += T::from(ch as u8 - b'0');
                }
                Some('e') => {
                    self.input = &self.input[1..];
                    return Ok(int);
                }
                _ => {
                    return Err(Error::ExpectedInteger);
                }
            }
        }
    }

    /// Parse a possible minus sign in front of an integer.
    /// Because in `1code` we have `-` sign between `i` and `e` it's easier to copy
    /// `parse_unsigned`.
    fn parse_signed<T>(&mut self) -> Result<T>
    where
        T: Neg<Output = T> + AddAssign<T> + MulAssign<T> + From<i8>,
    {
        if self.next_char()? != 'i' {
            return Err(Error::ExpectedInteger);
        }
        let mut negative = false;
        if self.peek_char()? == '-' {
            negative = true;
            self.input = &self.input[1..];
        }
        let mut int = T::from(0);
        loop {
            match self.input.chars().next() {
                Some(ch @ '0'..='9') => {
                    self.input = &self.input[1..];
                    int *= T::from(10);
                    int += T::from(ch as i8 - b'0' as i8);
                }
                Some('e') => {
                    self.input = &self.input[1..];
                    if negative {
                        return Ok(int.neg());
                    } else {
                        return Ok(int);
                    }
                }
                _ => {
                    return Err(Error::ExpectedInteger);
                }
            }
        }
    }

    /// Escape symblos are not supported.
    /// | rust  | 1coded |
    /// | ----- | ------ |
    /// | ""    | 0:     |
    /// | "0.1" | 3:0.1  |
    /// | "h h" | 3:h h  |
    fn parse_string(&mut self) -> Result<&'de str> {
        //TODO: rewrite to a logic - check that first char is int
        //self.input.find(':')
        //use result as length of length's int
        //parse length - take string
        let mut length = match self.next_char()? {
            ch @ '0'..='9' => usize::from(ch as u8 - b'0'),
            _ => {
                return Err(Error::ExpectedString);
            }
        };
        loop {
            match self.input.chars().next() {
                Some(ch @ '0'..='9') => {
                    self.input = &self.input[1..];
                    length *= usize::from(10 as u8);
                    length += usize::from(ch as u8 - b'0');
                }
                Some(':') => {
                    self.input = &self.input[1..];
                    break;
                }
                _ => {
                    return Err(Error::ExpectedString);
                }
            }
        }
        let string = &self.input[..length];
        self.input = &self.input[length..];
        return Ok(string);
    }
}

impl<'de, 'a> de::Deserializer<'de> for &'a mut Deserializer<'de> {
    type Error = Error;

    /// Because `1code` is self-describing format we can support `deserialize_any`.
    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        match self.peek_char()? {
            'N' => self.deserialize_unit(visitor),
            't' | 'f' => self.deserialize_bool(visitor),
            '0'..='9' => self.deserialize_str(visitor),
            'i' => {
                let mut iterator = self.input.chars();
                let _first_element = iterator.next().ok_or(Error::Eof)?;
                match iterator.next().ok_or(Error::Eof)? {
                    '-' => self.deserialize_i64(visitor),
                    '0'..='9' => self.deserialize_u64(visitor),
                    _ => Err(Error::Syntax),
                }
            }
            'l' => self.deserialize_seq(visitor),
            'd' => self.deserialize_map(visitor),
            _ => Err(Error::Syntax),
        }
    }

    fn deserialize_bool<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        visitor.visit_bool(self.parse_bool()?)
    }

    fn deserialize_i8<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        visitor.visit_i8(self.parse_signed()?)
    }

    fn deserialize_i16<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        visitor.visit_i16(self.parse_signed()?)
    }

    fn deserialize_i32<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        visitor.visit_i32(self.parse_signed()?)
    }

    fn deserialize_i64<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        visitor.visit_i64(self.parse_signed()?)
    }

    fn deserialize_u8<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        visitor.visit_u8(self.parse_unsigned()?)
    }

    fn deserialize_u16<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        visitor.visit_u16(self.parse_unsigned()?)
    }

    fn deserialize_u32<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        visitor.visit_u32(self.parse_unsigned()?)
    }

    fn deserialize_u64<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        visitor.visit_u64(self.parse_unsigned()?)
    }

    fn deserialize_f32<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_f64<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_char<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        let string = self.parse_string()?;
        if string.len() != 1 {
            return Err(Error::Syntax);
        } else {
            visitor.visit_char(
                string
                    .chars()
                    .next()
                    .expect("Next char should be presented."),
            )
        }
    }

    fn deserialize_str<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        visitor.visit_borrowed_str(self.parse_string()?)
    }

    fn deserialize_string<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        self.deserialize_str(visitor)
    }

    fn deserialize_bytes<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_byte_buf<V>(self, _visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        unimplemented!()
    }

    fn deserialize_option<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        if self.input.starts_with("N") {
            self.input = &self.input["N".len()..];
            visitor.visit_none()
        } else {
            visitor.visit_some(self)
        }
    }

    fn deserialize_unit<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        if self.input.starts_with("N") {
            self.input = &self.input["N".len()..];
            visitor.visit_unit()
        } else {
            Err(Error::ExpectedNull)
        }
    }

    fn deserialize_unit_struct<V>(self, _name: &'static str, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        self.deserialize_unit(visitor)
    }

    fn deserialize_newtype_struct<V>(self, _name: &'static str, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        visitor.visit_newtype_struct(self)
    }

    fn deserialize_seq<V>(mut self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        if self.next_char()? == 'l' {
            let value = visitor.visit_seq(NotSeparated::new(&mut self))?;
            if self.next_char()? == 'e' {
                Ok(value)
            } else {
                Err(Error::ExpectedListEnd)
            }
        } else {
            Err(Error::ExpectedList)
        }
    }

    fn deserialize_tuple<V>(self, _len: usize, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        self.deserialize_seq(visitor)
    }

    fn deserialize_tuple_struct<V>(
        self,
        _name: &'static str,
        _len: usize,
        visitor: V,
    ) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        self.deserialize_seq(visitor)
    }

    fn deserialize_map<V>(mut self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        if self.next_char()? == 'd' {
            let value = visitor.visit_map(NotSeparated::new(&mut self))?;
            if self.next_char()? == 'e' {
                Ok(value)
            } else {
                Err(Error::ExpectedDictionaryEnd)
            }
        } else {
            Err(Error::ExpectedDictionary)
        }
    }

    fn deserialize_struct<V>(
        self,
        _name: &'static str,
        _fields: &'static [&'static str],
        visitor: V,
    ) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        self.deserialize_map(visitor)
    }

    fn deserialize_enum<V>(
        self,
        _name: &'static str,
        _variants: &'static [&'static str],
        visitor: V,
    ) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        if self.next_char()? == 'd' {
            let value = visitor.visit_enum(Enum::new(self))?;
            if self.next_char()? == 'e' {
                Ok(value)
            } else {
                Err(Error::ExpectedDictionaryEnd)
            }
        } else {
            Err(Error::ExpectedDictionary)
        }
    }

    fn deserialize_identifier<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        self.deserialize_str(visitor)
    }

    fn deserialize_ignored_any<V>(self, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        self.deserialize_any(visitor)
    }
}

struct NotSeparated<'a, 'de: 'a> {
    de: &'a mut Deserializer<'de>,
}

impl<'a, 'de> NotSeparated<'a, 'de> {
    fn new(de: &'a mut Deserializer<'de>) -> Self {
        NotSeparated { de }
    }
}

impl<'de, 'a> SeqAccess<'de> for NotSeparated<'a, 'de> {
    type Error = Error;

    fn next_element_seed<T>(&mut self, seed: T) -> Result<Option<T::Value>>
    where
        T: DeserializeSeed<'de>,
    {
        if self.de.peek_char()? == 'e' {
            return Ok(None);
        } else {
            seed.deserialize(&mut *self.de).map(Some)
        }
    }
}

impl<'de, 'a> MapAccess<'de> for NotSeparated<'a, 'de> {
    type Error = Error;

    fn next_key_seed<K>(&mut self, seed: K) -> Result<Option<K::Value>>
    where
        K: DeserializeSeed<'de>,
    {
        //TODO: How we should understand is it end of the map or something inside?
        //Maybe we can hold a counter inside `NotSeparated` and
        //increment it on open chars (`i`, `l`, `d`) while decrement till 0.
        //But maybe serde will parse and consume all previous `e`?
        if self.de.peek_char()? == 'e' {
            return Ok(None);
        }
        seed.deserialize(&mut *self.de).map(Some)
    }

    fn next_value_seed<V>(&mut self, seed: V) -> Result<V::Value>
    where
        V: DeserializeSeed<'de>,
    {
        seed.deserialize(&mut *self.de)
    }
}

struct Enum<'a, 'de: 'a> {
    de: &'a mut Deserializer<'de>,
}

impl<'a, 'de> Enum<'a, 'de> {
    fn new(de: &'a mut Deserializer<'de>) -> Self {
        Enum { de }
    }
}

impl<'de, 'a> EnumAccess<'de> for Enum<'a, 'de> {
    type Error = Error;
    type Variant = Self;

    fn variant_seed<V>(self, seed: V) -> Result<(V::Value, Self::Variant)>
    where
        V: DeserializeSeed<'de>,
    {
        // The `deserialize_enum` method parsed a `d` character so we are
        // currently inside of a dictionary. The seed will be deserializing itself from
        // the key of the dictionary.
        let val = seed.deserialize(&mut *self.de)?;
        Ok((val, self))
    }
}

impl<'de, 'a> VariantAccess<'de> for Enum<'a, 'de> {
    type Error = Error;

    fn unit_variant(self) -> Result<()> {
        Err(Error::ExpectedString)
    }

    fn newtype_variant_seed<T>(self, seed: T) -> Result<T::Value>
    where
        T: DeserializeSeed<'de>,
    {
        seed.deserialize(self.de)
    }

    fn tuple_variant<V>(self, _len: usize, visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        de::Deserializer::deserialize_seq(self.de, visitor)
    }

    fn struct_variant<V>(self, _fields: &'static [&'static str], visitor: V) -> Result<V::Value>
    where
        V: Visitor<'de>,
    {
        de::Deserializer::deserialize_map(self.de, visitor)
    }
}

#[test]
fn test_int() {
    let test_1code = r#"i1e"#;
    let expected = 1;
    assert_eq!(expected, from_str(test_1code).unwrap());
}

#[test]
fn test_string() {
    let test_1code = r#"5:hello"#;
    let expected = "hello".to_string();
    let actual: String = from_str(test_1code).unwrap();
    assert_eq!(expected, actual);
}

#[test]
fn test_seq() {
    let test_1code = r#"l1:a1:be"#;
    let expected: Vec<String> = vec!["a".to_owned(), "b".to_owned()];
    let actual: Vec<String> = from_str(test_1code).unwrap();
    assert_eq!(expected, actual);
}

#[test]
fn test_struct() {
    #[derive(Deserialize, PartialEq, Debug)]
    struct Test {
        int: u32,
    }
    let test_1code = r#"d3:inti1ee"#;
    let expected = Test { int: 1 };
    assert_eq!(expected, from_str(test_1code).unwrap());
}

#[test]
fn test_struct_with_seq() {
    #[derive(Deserialize, PartialEq, Debug)]
    struct Test {
        int: u32,
        seq: Vec<String>,
    }

    let test_1code = r#"d3:inti1e3:seql1:a1:bee"#;
    let expected = Test {
        int: 1,
        seq: vec!["a".to_owned(), "b".to_owned()],
    };
    assert_eq!(expected, from_str(test_1code).unwrap());
}

#[test]
//TODO: should agree on https://github.com/soramitsu/1code-java/issues/26 first
#[ignore]
fn test_enum() {
    #[derive(Deserialize, PartialEq, Debug)]
    enum E {
        Unit,
        Newtype(u32),
        Tuple(u32, u32),
        Struct { a: u32 },
    }

    let j = r#""Unit""#;
    let expected = E::Unit;
    assert_eq!(expected, from_str(j).unwrap());

    let j = r#"{"Newtype":1}"#;
    let expected = E::Newtype(1);
    assert_eq!(expected, from_str(j).unwrap());

    let j = r#"{"Tuple":[1,2]}"#;
    let expected = E::Tuple(1, 2);
    assert_eq!(expected, from_str(j).unwrap());

    let j = r#"{"Struct":{"a":1}}"#;
    let expected = E::Struct { a: 1 };
    assert_eq!(expected, from_str(j).unwrap());
}


src/error.rs
use std;
use std::fmt::{self, Display};

use serde::{de, ser};

pub type Result<T> = std::result::Result<T, Error>;

#[derive(Clone, Debug, PartialEq)]
pub enum Error {
    Message(String),

    Eof,
    Syntax,
    ExpectedNull,
    ExpectedBoolean,
    ExpectedInteger,
    ExpectedString,
    ExpectedList,
    ExpectedListEnd,
    ExpectedDictionary,
    ExpectedDictionaryEnd,
    TrailingCharacters,
}

impl ser::Error for Error {
    fn custom<T: Display>(msg: T) -> Self {
        Error::Message(msg.to_string())
    }
}

impl de::Error for Error {
    fn custom<T: Display>(msg: T) -> Self {
        Error::Message(msg.to_string())
    }
}

impl Display for Error {
    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_str(std::error::Error::description(self))
    }
}

impl std::error::Error for Error {
    fn description(&self) -> &str {
        match *self {
            Error::Message(ref msg) => msg,
            Error::Eof => "unexpected end of input",
            //TODO: cover all variants
            _ => "uncovered case",
        }
    }
}


src/ser.rs
use crate::error::{Error, Result};
use serde::{ser, Serialize};

pub struct Serializer {
    output: String,
}

pub fn to_string<T>(value: &T) -> Result<String>
where
    T: Serialize,
{
    let mut serializer = Serializer {
        output: String::new(),
    };
    value.serialize(&mut serializer)?;
    Ok(serializer.output)
}

impl<'a> ser::Serializer for &'a mut Serializer {
    type Ok = ();
    type Error = Error;
    type SerializeSeq = Self;
    type SerializeTuple = Self;
    type SerializeTupleStruct = Self;
    type SerializeTupleVariant = Self;
    type SerializeMap = Self;
    type SerializeStruct = Self;
    type SerializeStructVariant = Self;

    /// | rust  | 1coded |
    /// | ----- | ------ |
    /// | true  |   T    |
    /// | false |   F    |
    fn serialize_bool(self, value: bool) -> Result<()> {
        self.output += if value { "T" } else { "F" };
        Ok(())
    }

    /// | rust  | 1coded |
    /// | ----- | ------ |
    /// | 0     | i0e    |
    /// | 42    | i42e   |
    /// | -1    | i-1e   |
    /// | 1.5   | i1.5e  |
    fn serialize_i64(self, value: i64) -> Result<()> {
        //TODO: replace with usage of https://crates.io/crates/itoa
        self.output += &format!("i{}e", value.to_string());
        Ok(())
    }

    fn serialize_i32(self, value: i32) -> Result<()> {
        self.serialize_i64(i64::from(value))
    }

    fn serialize_i16(self, value: i16) -> Result<()> {
        self.serialize_i64(i64::from(value))
    }

    fn serialize_i8(self, value: i8) -> Result<()> {
        self.serialize_i64(i64::from(value))
    }

    /// | rust  | 1coded |
    /// | ----- | ------ |
    /// | 0     | i0e    |
    /// | 42    | i42e   |
    /// | -1    | i-1e   |
    /// | 1.5   | i1.5e  |
    fn serialize_u64(self, value: u64) -> Result<()> {
        //TODO: replace with usage of https://crates.io/crates/itoa
        self.output += &format!("i{}e", value.to_string());
        Ok(())
    }

    fn serialize_u32(self, value: u32) -> Result<()> {
        self.serialize_u64(u64::from(value))
    }

    fn serialize_u16(self, value: u16) -> Result<()> {
        self.serialize_u64(u64::from(value))
    }

    fn serialize_u8(self, value: u8) -> Result<()> {
        self.serialize_u64(u64::from(value))
    }

    /// | rust  | 1coded |
    /// | ----- | ------ |
    /// | 0     | i0e    |
    /// | 42    | i42e   |
    /// | -1    | i-1e   |
    /// | 1.5   | i1.5e  |
    fn serialize_f64(self, value: f64) -> Result<()> {
        //TODO: replace with usage of https://crates.io/crates/itoa
        self.output += &format!("i{}e", value.to_string());
        Ok(())
    }

    fn serialize_f32(self, value: f32) -> Result<()> {
        self.serialize_f64(f64::from(value))
    }

    /// | rust  | 1coded |
    /// | ----- | ------ |
    /// | ""    | 0:     |
    /// | "0.1" | 3:0.1  |
    /// | "h h" | 3:h h  |
    fn serialize_str(self, value: &str) -> Result<()> {
        self.output += &format!("{}:{}", value.len(), value);
        Ok(())
    }

    fn serialize_char(self, value: char) -> Result<()> {
        self.serialize_str(&value.to_string())
    }

    /// |    rust     |    1coded   |
    /// | ----------- | ----------- |
    /// | vec![1,2,3] | li1ei2ei3ee |
    fn serialize_bytes(self, value: &[u8]) -> Result<()> {
        use serde::ser::SerializeSeq;
        let mut seq = self.serialize_seq(Some(value.len()))?;
        for byte in value {
            seq.serialize_element(byte)?;
        }
        seq.end()
    }

    /// List serialization: start.
    fn serialize_seq(self, _length: Option<usize>) -> Result<Self::SerializeSeq> {
        self.output += "l";
        Ok(self)
    }

    fn serialize_tuple(self, length: usize) -> Result<Self::SerializeTuple> {
        self.serialize_seq(Some(length))
    }

    fn serialize_tuple_struct(
        self,
        _name: &'static str,
        length: usize,
    ) -> Result<Self::SerializeTupleStruct> {
        self.serialize_seq(Some(length))
    }

    /// |       rust       |     1coded   |
    /// | ---------------- | ------------ |
    /// | { "str": "str" } | d3:str3:stre |
    fn serialize_newtype_variant<T>(
        self,
        _name: &'static str,
        _variant_index: u32,
        variant: &'static str,
        value: &T,
    ) -> Result<()>
    where
        T: ?Sized + Serialize,
    {
        self.output += "d";
        variant.serialize(&mut *self)?;
        value.serialize(&mut *self)?;
        self.output += "e";
        Ok(())
    }

    fn serialize_map(self, _len: Option<usize>) -> Result<Self::SerializeMap> {
        self.output += "d";
        Ok(self)
    }

    fn serialize_struct(self, _name: &'static str, len: usize) -> Result<Self::SerializeStruct> {
        self.serialize_map(Some(len))
    }

    fn serialize_struct_variant(
        self,
        _name: &'static str,
        _variant_index: u32,
        variant: &'static str,
        _len: usize,
    ) -> Result<Self::SerializeStructVariant> {
        self.output += "d";
        variant.serialize(&mut *self)?;
        self.output += "d";
        Ok(self)
    }

    fn serialize_tuple_variant(
        self,
        _name: &'static str,
        _variant_index: u32,
        variant: &'static str,
        _len: usize,
    ) -> Result<Self::SerializeTupleVariant> {
        self.output += "d";
        variant.serialize(&mut *self)?;
        self.output += "l";
        Ok(self)
    }

    /// |     rust     | 1coded |
    /// | ------------ | ------ |
    /// | ()           | N      |
    /// | Option::None | N      |
    fn serialize_unit(self) -> Result<()> {
        self.output += "N";
        Ok(())
    }

    fn serialize_none(self) -> Result<()> {
        self.serialize_unit()
    }

    fn serialize_some<T>(self, value: &T) -> Result<()>
    where
        T: ?Sized + Serialize,
    {
        value.serialize(self)
    }

    fn serialize_unit_struct(self, _name: &'static str) -> Result<()> {
        self.serialize_unit()
    }

    fn serialize_unit_variant(
        self,
        _name: &'static str,
        _variant_index: u32,
        variant: &'static str,
    ) -> Result<()> {
        //TODO decide https://github.com/soramitsu/1code-java/issues/26
        self.serialize_str(variant)
    }

    fn serialize_newtype_struct<T>(self, _name: &'static str, value: &T) -> Result<()>
    where
        T: ?Sized + Serialize,
    {
        value.serialize(self)
    }
}

impl<'a> ser::SerializeSeq for &'a mut Serializer {
    type Ok = ();
    type Error = Error;

    fn serialize_element<T>(&mut self, value: &T) -> Result<()>
    where
        T: ?Sized + Serialize,
    {
        value.serialize(&mut **self)
    }

    fn end(self) -> Result<()> {
        self.output += "e";
        Ok(())
    }
}

impl<'a> ser::SerializeTuple for &'a mut Serializer {
    type Ok = ();
    type Error = Error;

    fn serialize_element<T>(&mut self, value: &T) -> Result<()>
    where
        T: ?Sized + Serialize,
    {
        value.serialize(&mut **self)
    }

    fn end(self) -> Result<()> {
        self.output += "e";
        Ok(())
    }
}

impl<'a> ser::SerializeTupleStruct for &'a mut Serializer {
    type Ok = ();
    type Error = Error;

    fn serialize_field<T>(&mut self, value: &T) -> Result<()>
    where
        T: ?Sized + Serialize,
    {
        value.serialize(&mut **self)
    }

    fn end(self) -> Result<()> {
        self.output += "e";
        Ok(())
    }
}

impl<'a> ser::SerializeTupleVariant for &'a mut Serializer {
    type Ok = ();
    type Error = Error;

    fn serialize_field<T>(&mut self, value: &T) -> Result<()>
    where
        T: ?Sized + Serialize,
    {
        value.serialize(&mut **self)
    }

    fn end(self) -> Result<()> {
        self.output += "ee";
        Ok(())
    }
}

impl<'a> ser::SerializeMap for &'a mut Serializer {
    type Ok = ();
    type Error = Error;

    fn serialize_key<T>(&mut self, key: &T) -> Result<()>
    where
        T: ?Sized + Serialize,
    {
        key.serialize(&mut **self)
    }

    fn serialize_value<T>(&mut self, value: &T) -> Result<()>
    where
        T: ?Sized + Serialize,
    {
        value.serialize(&mut **self)
    }

    fn end(self) -> Result<()> {
        self.output += "e";
        Ok(())
    }
}

impl<'a> ser::SerializeStruct for &'a mut Serializer {
    type Ok = ();
    type Error = Error;

    fn serialize_field<T>(&mut self, key: &'static str, value: &T) -> Result<()>
    where
        T: ?Sized + Serialize,
    {
        key.serialize(&mut **self)?;
        value.serialize(&mut **self)
    }

    fn end(self) -> Result<()> {
        self.output += "e";
        Ok(())
    }
}

impl<'a> ser::SerializeStructVariant for &'a mut Serializer {
    type Ok = ();
    type Error = Error;

    fn serialize_field<T>(&mut self, key: &'static str, value: &T) -> Result<()>
    where
        T: ?Sized + Serialize,
    {
        key.serialize(&mut **self)?;
        value.serialize(&mut **self)
    }

    fn end(self) -> Result<()> {
        self.output += "ee";
        Ok(())
    }
}

use std::collections::HashMap;
#[derive(serde::Serialize)]
struct StructToSerialize {
    boolean: bool,
    positive_integer: u8,
    negative_integer: i16,
    positive_float: f32,
    negative_float: f32,
    negative_float_comma: f32,
    empty_string: String,
    number_string: String,
    latin_string: String,
    cyrillic_string: String,
    japan_string: String,
    number_list: Vec<u16>,
    string_list: Vec<String>,
    number_dictionary: HashMap<String, u16>,
    string_dictionary: HashMap<String, String>,
    list_dictionary: HashMap<String, Vec<u16>>,
    null: Option<()>,
}

// Transforms struct into
/*
    d
    7:booleanT
    16:positive_integeri1e
    16:negative_integeri-1e
    14:positive_floati1.5e
    14:negative_floati-1.5e
    20:negative_float_commai-1.5e
    12:empty_string0:
    13:number_string3:0.1
    12:latin_string11:hello world
    15:cyrillic_string19:привет мир
    12:japan_string21:こんにちは世界
    11:number_listli1ei2ei3ee
    11:string_listl1:12:013:011e
    17:number_dictionaryd1:1i1ee
    17:string_dictionaryd1:11:1e
    15:list_dictionaryd1:1li1ei2ei3eee
    4:nullN
    e
*/
#[test]
fn it_works() {
    let mut number_dictionary = HashMap::new();
    number_dictionary.insert("1".to_string(), 1);
    let mut string_dictionary = HashMap::new();
    string_dictionary.insert("1".to_string(), "1".to_string());
    let mut list_dictionary = HashMap::new();
    list_dictionary.insert("1".to_string(), vec![1, 2, 3]);
    let struct_to_serialize = StructToSerialize {
        boolean: true,
        positive_integer: 1,
        negative_integer: -1,
        positive_float: 1.5,
        negative_float: -1.5,
        negative_float_comma: -1.5,
        empty_string: "".to_string(),
        number_string: "0.1".to_string(),
        latin_string: "hello world".to_string(),
        cyrillic_string: "привет мир".to_string(),
        japan_string: "こんにちは世界".to_string(),
        number_list: vec![1, 2, 3],
        string_list: vec!["1".to_string(), "01".to_string(), "011".to_string()],
        number_dictionary: number_dictionary,
        string_dictionary: string_dictionary,
        list_dictionary: list_dictionary,
        null: Option::None,
    };
    let expected_result = "d7:booleanT16:positive_integeri1e16:negative_integeri-1e14:positive_floati1.5e14:negative_floati-1.5e20:negative_float_commai-1.5e12:empty_string0:13:number_string3:0.112:latin_string11:hello world15:cyrillic_string19:привет мир12:japan_string21:こんにちは世界11:number_listli1ei2ei3ee11:string_listl1:12:013:011e17:number_dictionaryd1:1i1ee17:string_dictionaryd1:11:1e15:list_dictionaryd1:1li1ei2ei3eee4:nullNe";
    assert_eq!(to_string(&struct_to_serialize).unwrap(), expected_result);
}
